torch/_functorch/aot_autograd.py

```diff
@@ -3357,6 +3655,264 @@ def forward(*runtime_args):
+            #     any pytrees in the graph.
+            # (2) We are guaranteed in the aot_export_module case that the forward outputs a loss,
+            #     and there are therefore no tangents that are needed to run the joint graph.
+            # (3) AOTAutograd creates a grad_input for every input in the forward,
+            #     including None's for inputs that are not grad-requiring tensors.
+            #     we don't want these in our export graph.
+            #     and there are therefore no tangents that are needed to run the joint graph.
+            # This function "fixes" both of the above by removing any tangent inputs,
+            # and removing pytrees from the original FX graph.
+            fake_tangents = [None for _ in range(metadata.num_outputs + metadata.num_mutated_inputs)]
+            fw_outs, gradients = fx_g(args, fake_tangents)
+            assert len(gradients) == len(args)
+            output_gradients = []
+            for i, (a, grad) in enumerate(zip(args, gradients)):
+                if isinstance(a, torch.Tensor) and a.requires_grad:
+                    assert grad is not None, """\
+Found a parameter that did not receive a gradient.
+"This is most likely a bug, but if this needs to be supported please file an issue on GitHub."""
+                    output_gradients.append(grad)
+                else:
+                    assert grad is None
+            return *fw_outs, *output_gradients
+        fx_g = make_fx(flattened_joint)(*full_args)
+
+    user_args_flat, _ = pytree.tree_flatten(args)
+    return fx_g, create_graph_signature(
+        fx_g,
+        metadata,
+        in_spec,
+        out_spec,
+        user_args_flat=user_args_flat,
+        params_and_buffers_flat=params_and_buffers_flat,
+        param_names=list(named_parameters.keys()),
+        buffer_names=list(named_buffers.keys()),
+        trace_joint=trace_joint,
+        num_user_fw_outs=num_fw_outs,
+        loss_index=output_loss_index,
+    )
+
+"""
+A simplified version of export. Used by higher order operators.
+
+This function makes a high-level "no calling convention changes" guarantee:
+- If no inputs require grad (so we export an inference graph),
+  there are *no* calling convention change between the exported graph, and "func".
+- If at least one input requires grad (so we trace out and expot a joint fw-bw graph),
+  Then if you were partition the graph into a separate forward and backward graph,
+  The forward graph will have no calling convention changes compared to "func".
+
+The above also relies on some strong restrictions around which functions this API accepts:
+(1) `args` cannot contain any pytrees (they must have been pytree_flattened already)
+(2) `func` cannot mutate any inputs
+(3) The outputs of `func` cannot alias any inputs.
+
+Note: this function is only lightly tested today. It will probably be tested more heavily by higher order ops.
```

Comment:
nit: If you actually want to docblock it, the docblock goes in the function, not on top.